Providers/Profile.swift

```diff
@@ -898,55 +946,66 @@ public class BrowserProfile: Profile {
-                    log.info("Not syncing \(label); already syncing something.")
-                    return deferMaybe(AlreadySyncingError())
-                }
-
-                if let label = label {
-                    log.info("Syncing \(label).")
-                }
-
-                let authState = account.syncAuthState
-
-                let readyDeferred = SyncStateMachine(prefs: self.prefsForSync).toReady(authState)
-                let delegate = profile.getSyncDelegate()
-
-                let go = readyDeferred >>== self.takeActionsOnEngineStateChanges >>== { ready in
-                    function(delegate, self.prefsForSync, ready)
-                }
-
-                // Always unlock when we're done.
-                go.upon({ res in self.endSyncing() })
-
-                return go
+        private func sync(label: EngineIdentifier, function: SyncFunction) -> SyncResult {
+            return syncSeveral((label, function)) >>== { statuses in
+                let engineStatus = statuses.filter({ $0.0 == label }).first!
+                return deferMaybe(engineStatus.1)
             }
-
-            log.warning("No account; can't sync.")
-            return nil
         }
 
         /**
-         * Runs the single provided synchronization function and returns its status.
+         * Runs each of the provided synchronization functions with the same inputs.
+         * Returns an array of IDs and SyncStatuses the same length as the input.
          */
-        private func sync(label: EngineIdentifier, function: (SyncDelegate, Prefs, Ready) -> SyncResult) -> SyncResult {
-            return self.withSyncInputs(label, function: function) ??
-                   deferMaybe(.NotStarted(.NoAccount))
+        private func syncSeveral(synchronizers: (EngineIdentifier, SyncFunction)...) -> Deferred<Maybe<[EngineStatus]>> {
+            return syncSeveral(synchronizers)
         }
 
         /**
          * Runs each of the provided synchronization functions with the same inputs.
          * Returns an array of IDs and SyncStatuses the same length as the input.
          */
-        private func syncSeveral(synchronizers: (EngineIdentifier, SyncFunction)...) -> Deferred<Maybe<[(EngineIdentifier, SyncStatus)]>> {
-            typealias Pair = (EngineIdentifier, SyncStatus)
-            let combined: (SyncDelegate, Prefs, Ready) -> Deferred<Maybe<[Pair]>> = { delegate, syncPrefs, ready in
+        private func syncSeveral(synchronizers: [(EngineIdentifier, SyncFunction)]) -> Deferred<Maybe<[EngineStatus]>> {
+
+            syncLock.lock()
+            defer { syncLock.unlock() }
+            let go: Deferred<Maybe<[EngineStatus]>>
+            let requestedLabels = synchronizers.map { $0.0 }
+            if !beginSyncing() {
+                log.info("Already syncing something. Will queue for later: \(requestedLabels)")
+                go = currentSync! >>== { justSynced in self.syncRemaining(synchronizers, except: justSynced) }
```

Comment:
I'm not a big fan of this combination of `beginSyncing` and `currentSync!` — indeed, I'm not a big fan of `beginSyncing` at all. Can we shift that side-effect into a `didSet` method on one of the `Deferred` slots?